### PR TITLE
Implement shared memory flag w/ encoding/decoding

### PR DESCRIPTION
**Pull request by binji**
This also enforces requirement that all atomic instructions have shared
memory.

### CODE REVIEW

File: interpreter/valid/valid.ml [hunk 1]
```diff
@@ -155,6 +155,12 @@ let check_memop (c : context) (memop : 'a memop) get_sz at =
   require (1 lsl memop.align <= size) at
     "alignment must not be larger than natural"
 
+let check_atomic_memop (c : context) (memop : 'a memop) get_sz at =
```
Comment (rossberg) [inline interpreter/valid/valid.ml]: How about keeping the single `check_memop` and pass it an optional expected sharability `sh` and then check"
```
require (sh = None || sh = Some share) ...
```

File: interpreter/binary/decode.ml [hunk 2]
```diff
@@ -96,7 +96,7 @@ let rec vsN n s =
   then (if b land 0x40 = 0 then x else Int64.(logor x (logxor (-1L) 0x7fL)))
   else Int64.(logor x (shift_left (vsN (n - 7) s) 7))
 
-let vu1 s = Int64.to_int (vuN 1 s)
+let vu2 s = Int64.to_int (vuN 2 s)
```
Comment (rossberg) [inline interpreter/binary/decode.ml]: Perhaps we should just eliminate this and use u8 below.

File: interpreter/text/parser.mly [hunk 3]
```diff
@@ -235,7 +235,8 @@ table_sig :
   | limits elem_type { TableType ($1, $2) }
 
 memory_sig :
-  | limits { MemoryType $1 }
+  | limits { MemoryType ($1, Unshared) }
+  | LPAR SHARED limits RPAR { MemoryType ($3, Shared) }
```
[the remark this answers is not in the document]
Comment (binji) [reply interpreter/text/parser.mly]: Hm, I figured it was more like mutable globals.

File: interpreter/binary/decode.ml (hunk 2 above)
Comment (binji) [reply interpreter/binary/decode.ml]: Could do, but then would have to check that high bits are zero in callers, which is a little inconvenient.

File: interpreter/binary/decode.ml [hunk 4]
```diff
@@ -110,7 +110,7 @@ let len32 s =
   if n <= Int32.of_int (len s) then Int32.to_int n else
     error s pos "length out of bounds"
 
-let bool s = (vu1 s = 1)
+let bool2 s = let n = vu2 s in ((n land 2) = 2), ((n land 1) = 1)
```
[the remark this answers is not in the document]
Comment (binji) [reply interpreter/binary/decode.ml]: Done

File: interpreter/binary/decode.ml [hunk 5]
```diff
@@ -160,19 +160,19 @@ let func_type s =
...
+  let has_max, flag = bool2 s in
   let min = vu s in
   let max = opt vu has_max s in
-  {min; max}
+  {min; max}, flag
 
 let table_type s =
   let t = elem_type s in
-  let lim = limits vu32 s in
+  let lim, _ = limits vu32 s in
```
[the remark this answers is not in the document]
Comment (binji) [reply interpreter/binary/decode.ml]: Done

File: interpreter/binary/decode.ml [hunk 6]
```diff
@@ -160,19 +160,19 @@ let func_type s =
...
   TableType (lim, t)
 
 let memory_type s =
-  let lim = limits vu32 s in
-  MemoryType lim
+  let lim, share = limits vu32 s in
+  MemoryType (lim, (if share then Shared else Unshared))
```
[the remark this answers is not in the document]
Comment (binji) [reply interpreter/binary/decode.ml]: Done

File: interpreter/syntax/types.ml [hunk 7]
```diff
@@ -39,8 +40,8 @@ let match_func_type ft1 ft2 =
 let match_table_type (TableType (lim1, et1)) (TableType (lim2, et2)) =
   et1 = et2 && match_limits lim1 lim2
 
-let match_memory_type (MemoryType lim1) (MemoryType lim2) =
-  match_limits lim1 lim2
+let match_memory_type (MemoryType (lim1, sh1)) (MemoryType (lim2, sh2)) =
+  sh1 == sh2 && match_limits lim1 lim2
```
[the remark this answers is not in the document]
Comment (binji) [reply interpreter/syntax/types.ml]: Done

File: interpreter/valid/valid.ml (hunk 1 above)
Comment (binji) [reply interpreter/valid/valid.ml]: Done

File: interpreter/host/spectest.ml [hunk 8]
```diff
@@ -17,7 +17,8 @@ let global (GlobalType (t, _) as gt) =
   in Global.alloc gt v
 
 let table = Table.alloc (TableType ({min = 10l; max = Some 20l}, AnyFuncType))
-let memory = Memory.alloc (MemoryType {min = 1l; max = Some 2l})
+let memory = Memory.alloc (MemoryType ({min = 1l; max = Some 2l}, Unshared))
+let shared_memory = Memory.alloc (MemoryType ({min = 1l; max = Some 2l}, Shared))
```
Comment (rossberg) [inline interpreter/host/spectest.ml]: Nit: line break (above as well while you're here)

File: interpreter/runtime/memory.ml [hunk 9]
```diff
@@ -11,7 +11,7 @@ type mem_size = Mem8 | Mem16 | Mem32
 type extension = SX | ZX
 
 type memory' = (int, int8_unsigned_elt, c_layout) Array1.t
-type memory = {mutable content : memory'; max : size option}
+type memory = {mutable content : memory'; max : size option; shared: sharability}
```
Comment (rossberg) [inline interpreter/runtime/memory.ml]: Nit: line break

File: interpreter/binary/decode.ml [hunk 10]
```diff
@@ -160,19 +160,20 @@ let func_type s =
...
-  let lim = limits vu32 s in
-  TableType (lim, t)
+  match limits vu32 s with
+  | lim, false -> TableType (lim, t)
+  | _, true -> error s (pos s - 1) "invalid table type"
 
 let memory_type s =
-  let lim = limits vu32 s in
-  MemoryType lim
+  let lim, share = limits vu32 s in
```
Comment (rossberg) [inline interpreter/binary/decode.ml]: Nit: `sh` or `shared`

File: interpreter/text/parser.mly (hunk 3 above)
[the remark this answers is not in the document]
Comment (rossberg) [reply interpreter/text/parser.mly]: The rationale is that mutability applies "pointwise" to the content (the syntax was intended to scale e.g. to structs, where fields may be im/mutable individually), while sharability applies to the whole thing.

File: interpreter/valid/valid.ml [hunk 11]
```diff
@@ -142,8 +142,8 @@ let check_unop unop at =
   | Values.I32 I32Op.Extend32S -> error at "invalid unary operator"
   | _ -> ()
 
-let check_memop (c : context) (memop : 'a memop) get_sz at =
-  ignore (memory c (0l @@ at));
+let check_memop (c : context) (memop : 'a memop) get_sz sh at =
+  let MemoryType (_, share) = memory c (0l @@ at) in
```
Comment (rossberg) [inline interpreter/valid/valid.ml]: Nit: shared

File: interpreter/valid/valid.ml [hunk 12]
```diff
@@ -153,7 +153,8 @@ let check_memop (c : context) (memop : 'a memop) get_sz at =
       Memory.mem_size sz
   in
   require (1 lsl memop.align <= size) at
-    "alignment must not be larger than natural"
+    "alignment must not be larger than natural";
+  require (sh = None || sh = Some share) at "atomic accesses require shared memory"
```
Comment (rossberg) [inline interpreter/valid/valid.ml]: Nit: line break

File: interpreter/valid/valid.ml [hunk 13]
```diff
@@ -142,8 +142,8 @@ let check_unop unop at =
   | Values.I32 I32Op.Extend32S -> error at "invalid unary operator"
   | _ -> ()
 
-let check_memop (c : context) (memop : 'a memop) get_sz at =
-  ignore (memory c (0l @@ at));
+let check_memop (c : context) (memop : 'a memop) get_sz sh at =
```
Comment (rossberg) [inline interpreter/valid/valid.ml]: Nit: can you put a type annotation on the `sh` parameter, and place it before `get_sz`? It is part of the judgement that this function implements, while the latter is just an auxiliary thing like `at`.

File: interpreter/valid/valid.ml [hunk 14]
```diff
@@ -354,9 +355,11 @@ let check_memory_size (sz : I32.t) at =
     "memory size must be at most 65536 pages (4GiB)"
 
 let check_memory_type (mt : memory_type) at =
-  let MemoryType lim = mt in
+  let MemoryType (lim, share) = mt in
```
Comment (rossberg) [inline interpreter/valid/valid.ml]: Nit: shared

File: interpreter/binary/encode.ml [hunk 15]
```diff
@@ -109,14 +108,14 @@ let encode m =
...
+    let limits vu f1 {min; max} =
+      bool2 (max <> None) f1; vu min; opt vu max
 
     let table_type = function
-      | TableType (lim, t) -> elem_type t; limits vu32 lim
+      | TableType (lim, t) -> elem_type t; limits vu32 false lim
 
     let memory_type = function
-      | MemoryType lim -> limits vu32 lim
+      | MemoryType (lim, share) -> limits vu32 (share = Shared) lim
```
Comment (rossberg) [inline interpreter/binary/encode.ml]: Nit: `sh` or `shared`

File: interpreter/binary/decode.ml (hunk 2 above)
Comment (binji) [reply interpreter/binary/decode.ml]: Done.

File: interpreter/text/parser.mly (hunk 3 above)
[the remark this answers is not in the document]
Comment (binji) [reply interpreter/text/parser.mly]: OK, done.

File: interpreter/binary/decode.ml [hunk 16]
```diff
@@ -160,19 +160,20 @@ let func_type s =
...
   let min = vu s in
   let max = opt vu has_max s in
-  {min; max}
+  {min; max}, flag
 
 let table_type s =
   let t = elem_type s in
-  let lim = limits vu32 s in
-  TableType (lim, t)
+  match limits vu32 s with
```
[the remark this answers is not in the document]
Comment (binji) [reply interpreter/binary/decode.ml]: Done.

File: interpreter/binary/decode.ml (hunk 10 above)
Comment (binji) [reply interpreter/binary/decode.ml]: Done.

File: interpreter/binary/encode.ml (hunk 15 above)
Comment (binji) [reply interpreter/binary/encode.ml]: Done.

File: interpreter/host/spectest.ml (hunk 8 above)
Comment (binji) [reply interpreter/host/spectest.ml]: Done. (which one do you mean above?)

File: interpreter/runtime/memory.ml (hunk 9 above)
Comment (binji) [reply interpreter/runtime/memory.ml]: Done.

File: interpreter/valid/valid.ml (hunk 13 above)
Comment (binji) [reply interpreter/valid/valid.ml]: Done.

File: interpreter/valid/valid.ml (hunk 11 above)
Comment (binji) [reply interpreter/valid/valid.ml]: Done.

File: interpreter/valid/valid.ml (hunk 12 above)
Comment (binji) [reply interpreter/valid/valid.ml]: Done.

File: interpreter/valid/valid.ml (hunk 14 above)
Comment (binji) [reply interpreter/valid/valid.ml]: Done.

File: interpreter/binary/decode.ml [hunk 17]
```diff
@@ -110,7 +109,10 @@ let len32 s =
   if n <= Int32.of_int (len s) then Int32.to_int n else
     error s pos "length out of bounds"
 
-let bool s = (vu1 s = 1)
+let bool2 s = let n = u8 s in
```
Comment (rossberg) [inline interpreter/binary/decode.ml]: Nit: line break before `let`

File: interpreter/syntax/types.ml [hunk 18]
```diff
@@ -86,7 +87,8 @@ let string_of_limits {min; max} =
   (match max with None -> "" | Some n -> " " ^ I32.to_string_u n)
 
 let string_of_memory_type = function
-  | MemoryType lim -> string_of_limits lim
+  | MemoryType (lim, Unshared) -> string_of_limits lim
+  | MemoryType (lim, Shared) -> "(shared " ^ string_of_limits lim ^ ")"
```
Comment (rossberg) [inline interpreter/syntax/types.ml]: Probably want to adjust that one too

File: interpreter/binary/decode.ml (hunk 17 above)
Comment (binji) [reply interpreter/binary/decode.ml]: Done

File: interpreter/syntax/types.ml (hunk 18 above)
Comment (binji) [reply interpreter/syntax/types.ml]: Done

**Pull request merged**